ledger/participant-integration-api/src/main/scala/platform/store/backend/common/EventReaderQueries.scala

```diff
@@ -0,0 +1,185 @@
+      ),
+      SelectTable(
+        tableName = "participant_events_consuming_exercise",
+        selectColumns =
+          s"$selectColumnsForFlatTransactionsExercise, ${queryStrategy.constBooleanSelect(true)} as exercise_consuming",
+      ),
+    )
+
+    def selectFrom(tableName: String, selectColumns: String) =
+      cSQL"""
+      (
+        SELECT
+          #$selectColumns,
+          event_witnesses,
+          command_id
+        FROM
+        (
+            SELECT
+              #$selectColumns,
+              #$witnessesColumn as event_witnesses,
+              command_id
+            FROM
+              #$tableName
+            WHERE
+              contract_id = ${contractId.coid}
+            ORDER BY
+              event_sequential_id
+        ) x
+      )
+    """
+
+    val unionQuery = tables
+      .map(table =>
+        selectFrom(
+          tableName = table.tableName,
+          selectColumns = table.selectColumns,
+        )
+      )
+      .mkComposite("", " UNION ALL", "")
+
+    val query = SQL"""$unionQuery ORDER BY event_sequential_id"""
+
+    query.asVectorOf(eventParser(requestingParties))(connection)
+
+  }
+
+  def resolveEventIdToSequentialId(
+      eventId: EventId
+  )(conn: Connection): Option[EventSequentialId] = {
+    val parser = SqlParser.get[EventSequentialId](1)
+    val query = SQL"""
+      SELECT c.event_sequential_id
+      FROM participant_transaction_meta m
+      JOIN participant_events_create c ON c.event_offset = m.event_offset
+      WHERE m.transaction_id = ${eventId.transactionId.toString}
+      AND c.event_id = ${eventId.toLedgerString.toString}
+    """
+    query.as(parser.singleOpt)(conn)
+  }
+
+  private def selectLatestKeyCreateEvent(
+      keyHash: String,
+      intRequestingParties: Set[Int],
+      extRequestingParties: Set[String],
+      lastExclusiveSeqId: EventSequentialId,
+  )(conn: Connection): Option[EventStorageBackend.Entry[FlatEvent.Created]] = {
+
+    @tailrec def go(
+        endExclusiveSeqId: EventSequentialId
+    ): Option[EventStorageBackend.Entry[FlatEvent.Created]] = {
+      val query =
+        SQL"""
+        WITH max_event AS (
+            SELECT max(c.event_sequential_id) AS sequential_id
+            FROM participant_events_create c
+            WHERE c.create_key_hash = $keyHash
+            AND c.event_sequential_id < $endExclusiveSeqId)
+        SELECT  #$selectColumnsForFlatTransactionsCreate,
+                #${queryStrategy.constBooleanSelect(false)} exercise_consuming,
+                flat_event_witnesses event_witnesses,
+                command_id
+        FROM max_event
+        JOIN participant_events_create c on c.event_sequential_id = max_event.sequential_id
+      """
+      query.as(createdFlatEventParser(intRequestingParties, stringInterning).singleOpt)(
+        conn
+      ) match {
+        case Some(c) if c.event.stakeholders.exists(extRequestingParties) => Some(c)
+        case Some(c) => go(c.eventSequentialId)
```

Comment:
I would recommend to have some upper bound of this iteration, otherwise this can easily spin out of control.